### DISCUSSION TITLE
Gradual migration path for vite's client/server separation?

### DISCUSSION THREAD

**Discussion by jrr** [remix-run/remix#8974] [Q&A]
I took a stab at migrating my project to Vite, and hit a hurdle splitting up client and server code, as described in the [migration guide](https://remix.run/docs/en/main/future/vite#splitting-up-client-and-server-code). The problem is that the necessary changes (whether renaming files or going wild with `serverOnly$()`) will produce a _large_ diff, and I can't reasonably land it all at once on a project with multiple active contributors and branches.

So what I'm after is a gentler migration path: ideally, opting in to the stricter client/server separation file-by-file, perhaps with a future flag in the old compiler, or maybe with the new Vite system in some kind of legacy mode.

Does something like this exist? Is it planned?

**Comment by jrestall** [comment]
What about doing the client and server code split changes before moving to Vite? For example moving all server-only code to *.server.ts files while still using esbuild. A longer path but a more incremental one.

Then you could have a vite feature branch and merge the changes as you go to monitor progress.

**Reply by jrr** [reply]
Yeah, that's sort of my plan "B":
 - keep a long-lived vite branch that will be broken for a while
 - periodically check on what the next broken piece is, then:
   - make the fix on the development branch
   - integrate those fixes back into the vite branch
   - (repeat)

One problem is that there's nothing stopping ongoing work on the development branch from digging the hole deeper. Maybe an ESLint rule could help developers avoid introducing more client/server separation problems?
